### ISSUE TITLE
pdnsutil add-record creates duplicate records

### ISSUE THREAD

**Issue by jagardaniel** [PowerDNS/pdns#4727]
- Program: pdnsutil
 - Issue type: Bug report

### Short description
pdnsutil add-record creates duplicate records

### Environment
 - Operating system: Ubuntu 16.04.1 LTS
 - Software version: pdnsutil 4.0.1
 - Software source: http://repo.powerdns.com/ubuntu

### Steps to reproduce
1. Create a zone
```
root@ns01:~# pdnsutil create-zone example.org
Creating empty zone 'example.org'
```
2. Create some records with the same value
```
root@ns01:~# pdnsutil add-record example.org test A 127.0.0.1
New rrset:
test.example.org. IN A 3600 127.0.0.1

root@ns01:~# pdnsutil add-record example.org test A 127.0.0.1
New rrset:
test.example.org. IN A 3600 127.0.0.1
test.example.org. IN A 3600 127.0.0.1

root@ns01:~# pdnsutil add-record example.org test A 127.0.0.1
New rrset:
test.example.org. IN A 3600 127.0.0.1
test.example.org. IN A 3600 127.0.0.1
test.example.org. IN A 3600 127.0.0.1
```

3. List zone
```
root@ns01:~# pdnsutil list-zone example.org
example.org	3600	IN	SOA	a.misconfigured.powerdns.server hostmaster.example.org 1 10800 3600 604800 3600
test.example.org	3600	IN	A	127.0.0.1
test.example.org	3600	IN	A	127.0.0.1
test.example.org	3600	IN	A	127.0.0.1
```

### Expected behaviour
I'm not exactly sure what the expected behaviour should be in this case. It feels a little bit redundant creating the same record again. Should it complain that the record already exists or just return ok without touching it?

### Usecase
I'm trying to use Ansible for creating zone and records in the zone. The playbook should be able to run multiple times and ensure that all the tasks are ok. If I'm using add-record right now Ansible will create a new record (with the same value) every time the playbook is executed. It is not really an issue since it still works, but it feels a little bit messy to have a lot of duplicate records in the zone.

**Comment by Habbie** [comment]
The expected behaviour is to not create the record. Error or quiet 'success' is a good question.

**Comment by jagardaniel** [comment]
Hi again! I would like to try creating a fix for this issue since it's not very critical and it could be a fun and (hopefully) easy task to start with.

I still don't know if it should throw an error message or just return "ok" when the same records already exist. I think return ok is the easiest way from a user perspective, and the only reason I can find for throwing an error message is if you want to know if the record you are trying to add already exists or being added for the first time. But this also means that the user can't just run add-record multiple times for the same record (like it would in a Ansible playbook) without having to ignore/catch when it happens the second time and all the times after that. What do you think?

**Comment by Habbie** [comment]
@jagardaniel 'ok' seems fine to me!

**Comment by Habbie** [comment]
This bug is still present on master, and thus in 4.2.0-rc1.

**Comment by snesbittsea** [comment]
With respect to status messages one consideration is Ansible idempotency. A return of ok would - I believe - make idempotency impossible - it would always show as changed.

Ergo, I wonder if attempt to add a duplicate shouldn't return an error with a result of "already exists".

**Comment by Habbie** [comment]
> With respect to status messages one consideration is Ansible idempotency. A return of ok would - I believe - make idempotency impossible - it would always show as changed.

A return of 'ok' would, in fact, make the call idempotent. The problem is that Ansible still wants to know whether 'it is as you want now' was a change or not. I understand and respect that, but pdnsutil is not meant to be a programmatic interface.

> Ergo, I wonder if attempt to add a duplicate shouldn't return an error with a result of "already exists".

That would pretty much be the opposite of idempotency!

> Edit: Hmm, I wonder if another approach would be to use the PowerDDNS API. Does anyone know what the API returns when a duplicate record is provided for insertion?

The only way to insert records via our REST API is to REPLACE a full RRset. Hence, there is no problem with duplicate records, and you also get idempotency (on our side, I am unsure we can usefully tell the client if anything -changed-). The REST API, or RFC2136 nsupdate, are the right interface for anything programmatic, such as Ansible.

**Comment by MEschenbacher** [comment]
I do not think this is a bug and I also do not think the expected behavour should be to not create the second identical RR.
If one chooses to create duplicate records, one should be able to do so.

The argument about idempotency is very important -- for me personally, too -- but I do not think the `pdnsutil add-record` command should guarantee idempotency because that is not what I expect from a low-level command for modifying zones.

If we're interested in a idempotent command, I'd suggest something along the lines of implementing a `pdnsutil ensure-record`.

**Comment by Habbie** [comment]
> If one chooses to create duplicate records, one should be able to do so because for the most cases, the result is a perfectly valid zone.

No, that's wrong. Duplicate records are strictly forbidden, and PowerDNS will not serve them anyway.

> If we're interested in a idempotent command, I'd suggest something along the lines of implementing a `pdnsutil ensure-record`.

`replace-rrset` is idempotent, too :)

**Comment by zeha** [comment]
> `replace-rrset` is idempotent, too :)

Given this, I'd think add-record should really fail with an error.

**Comment by MEschenbacher** [comment]
On 2021-05-13 03:34:48, Peter van Dijk wrote:
>> If one chooses to create duplicate records, one should be able to do
>> so because for the most cases, the result is a perfectly valid zone.
>
>No, that's wrong. Duplicate records are strictly forbidden, and
>PowerDNS will not serve them anyway.

Are you referring to
https://datatracker.ietf.org/doc/html/rfc2181#section-5 ?

I agree that there is no actual use for duplicate records and
`add-record` should indicate the error condition to the user.

>> If we're interested in a idempotent command, I'd suggest something
>> along the lines of implementing a `pdnsutil ensure-record`.
>
>`replace-rrset` is idempotent, too :)

`replace-rrset` works for replacing the entire rrset. It's definitely
useful, owever it does not replicate the exact semantics of
`ensure-record`.
